File: src/main/java/com/josephbaca/world/CoordinateGrid.java
package com.josephbaca.world;

import java.util.ArrayList;
import java.util.List;

/**
 * Easy to use class for representing things that follow a cartesian coordinate grid.
 */
public class CoordinateGrid<T extends Mappable> {

  private final List<List<T>> grid;

  CoordinateGrid(int x, int y) {
    grid = new ArrayList<>();
    for (int i = 0; i < x; i++) {
      grid.add(new ArrayList<>());
      for (int j = 0; j < y; j++) {
        grid.get(i).add(null);
      }
    }
  }

  /**
   * Gets the T at the provided coordinate.
   */
  public T getCoordinate(Coordinate c) {
    return grid.get(c.x).get(c.y);
  }

  /**
   * Sets a coordinate to a specified T.
   */
  public void setCoordinate(Coordinate c, T t) {
    grid.get(c.y).set(c.x, t);
  }

  @Override
  public String toString() {
    return grid.toString();
  }

  /**
   * Returns a print friendly version of the grid formatted as a standard cartesian coordinate grid.
   */
  String toDisplayString() {
    int sizeY = grid.size();
    int sizeX = sizeY != 0
              ? grid.get(0).size()
              : 0;

    StringBuilder sb = new StringBuilder(sizeX * sizeY);

    for (int i = sizeY - 1; i >= 0; i--) {
      for (int j = 0; j < sizeX; j++) {
        sb.append(getCoordinate(Coordinate.of(i, j)).getIcon());
        sb.append(" ");
      }
      sb.deleteCharAt(sb.length() - 1);
      sb.append("\n");
    }
    sb.deleteCharAt(sb.length() - 1);
    return sb.toString();
  }

  public int sizeX() {
    return grid.size();
  }

  public int sizeY() {
    return grid.get(0).size();
  }
}

File: src/main/java/com/josephbaca/world/CoordinateGrid.kt
package com.josephbaca.world

/**
 * Easy to use class for representing things that follow a cartesian coordinate grid.
 */
class CoordinateGrid<T : Mappable> internal constructor(val sizeX: Int, val sizeY: Int) {

    /**
     * Internal storage as y lists of x coords
     */
    private val grid: MutableList<MutableList<T?>> = MutableList(sizeY) { MutableList<T?>(sizeX) { null } }


    fun getCoordinate(c: Coordinate): T {
        return grid[c.y][c.x]!!
    }

    fun setCoordinate(c: Coordinate, t: T) {
        grid[c.y][c.x] = t
    }

    override fun toString(): String {
        return grid.toString()
    }

    /**
     * Returns a print friendly version of the grid formatted as a standard cartesian coordinate grid.
     */
    fun toDisplayString(): String {

        val sb = StringBuilder(sizeX * sizeY)

        for (j in sizeY - 1 downTo 0) { // Iterate down the Ys to order the list on a grid
            for (i in 0 until sizeX) {
                sb.append(getCoordinate(Coordinate(i, j)).icon)
                sb.append(" ")
            }
            sb.deleteCharAt(sb.length - 1)
            sb.append("\n")
        }
        sb.deleteCharAt(sb.length - 1)
        return sb.toString()
    }
}
